ping/main.go

```diff
@@ -0,0 +1,284 @@
+				continue // skip over ourselves.
+			}
+			peers = append(peers, ai)
+		case err := <-sub.Done():
+			return err
+		}
+	}
+	cancel()
+
+	// ✨
+	// ✨  Now we know about all other libp2p hosts in this test.
+	// ✨
+
+	// This is a closure that pings all peers in the test in parallel, and
+	// records the latency value as a message and as a result datapoint.
+	pingPeers := func(tag string) error {
+		g, gctx := errgroup.WithContext(ctx)
+		for _, ai := range peers {
+			ai := ai
+			g.Go(func() error {
+				// a context for the continuous stream of pings.
+				pctx, cancel := context.WithCancel(gctx)
+				defer cancel()
+				res := <-ping.Ping(pctx, ai.ID)
+				if res.Error != nil {
+					return res.Error
+				}
+
+				// record a message.
+				runenv.RecordMessage("ping result (%s) from peer %s: %s", tag, ai.ID, res.RTT)
+
+				// record a result point; these points will be batch-inserted
+				// into InfluxDB when the test concludes.
+				//
+				// ping-result is the metric name, and round and peer are tags.
+				point := fmt.Sprintf("ping-result,round=%s,peer=%s", tag, ai.ID)
+				runenv.R().RecordPoint(point, float64(res.RTT.Milliseconds()))
+				return nil
+			})
+		}
+		return g.Wait()
+	}
+
+	// ☎️  Connect to all other peers.
+	//
+	// Note: we sidestep simultaneous connect issues by ONLY connecting to peers
+	// who published their addresses before us (this is enough to dedup and avoid
+	// two peers dialling each other at the same time).
+	//
+	// We can do this because sync service pubsub is ordered.
+	for _, ai := range peers {
+		if ai.ID == id {
+			break
+		}
+		if err := host.Connect(ctx, *ai); err != nil {
+			return err
+		}
+	}
+
+	runenv.RecordMessage("done dialling my peers")
+
+	// Wait for all peers to signal that they're done with the connection phase.
+	initCtx.SyncClient.MustSignalAndWait(ctx, "connected", runenv.TestInstanceCount)
+
+	// 📡  Let's ping all our peers without any traffic shaping rules.
+	if err := pingPeers("initial"); err != nil {
+		return err
+	}
+
+	// 🕐  Wait for all peers to have finished the initial round.
+	initCtx.SyncClient.MustSignalAndWait(ctx, "initial", runenv.TestInstanceCount)
+
+	// 🎉 🎉 🎉
+	//
+	// Here is where the fun begins. We will perform `iterations` rounds of
+	// randomly altering our network latency, waiting for all other peers to
+	// do too. We will record our observations for each round.
+	//
+	// 🎉 🎉 🎉
+
+	// Let's initialize the random seed to the current timestamp + our global sequence number.
+	// Otherwise all instances will end up generating the same "random" latencies 🤦‍
+	rand.Seed(time.Now().UnixNano() + initCtx.GlobalSeq)
+
+	for i := 1; i <= iterations; i++ {
+		runenv.RecordMessage("⚡️  ITERATION ROUND %d", i)
+
+		// 🤹  Let's calculate our new latency.
+		latency := time.Duration(rand.Int31n(int32(maxLatencyMs))) * time.Millisecond
+		runenv.RecordMessage("(round %d) my latency: %s", i, latency)
+
+		// 🐌  Let's ask the NetClient to reconfigure our network.
```

Comment:
Do the UTF emojis help? No. Do they make the code more clear? No. Are they so god damn lovely and show the amount of dedication put into this document? Hell yes.

Reply:
Hahaha, my intention was for them to serve as markers of the large "phases" of the test.

Reply:
Glad they lightened your day.